src/PostgREST/Client.hs

```diff
@@ -0,0 +1,47 @@
+{-|
+Module      : PostgREST.Client
+Description : PostgREST HTTP client
+-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module PostgREST.Client
+  ( ready
+  ) where
+
+import qualified Data.Text           as T
+import qualified Network.HTTP.Client as HC
+
+import Network.HTTP.Client       (HttpException (..))
+import Network.HTTP.Types.Status (Status (..))
+import PostgREST.Config          (AppConfig (..))
+
+import Protolude
+
+ready :: AppConfig -> IO Status
+ready AppConfig{configAdminServerHost, configAdminServerPort} = do
+
+  client <- HC.newManager HC.defaultManagerSettings
+  req <- HC.parseRequest $ -- Create HTTP Request
+    case configAdminServerPort of
+      Just port -> "http://" <> host <> ":" <> show port <> "/ready"
+      Nothing   -> panic "Admin Server is not running. Please check your configuration."
+
+  resp <- HC.httpLbs req client `catch` \(_ :: HttpException) -> do
+    let url = show (HC.getUri req)
+    hPutStrLn stderr $ T.pack $ "postgrest: health check request failed - connection refused to " <> url
+    exitWith $ ExitFailure 1
+
+  return $ HC.responseStatus resp
+    where
+      host = escapeHostName $ T.unpack configAdminServerHost
+      -- We don't need to resolve the host name except the postgrest special
+      -- addresses. The http-client package automatically resolves the other
+      -- host names for us.
+      escapeHostName = \case
+        "*"  -> "0.0.0.0"
+        "*4" -> "0.0.0.0"
+        "!4" -> "0.0.0.0"
+        "*6" -> "[::1]" -- TODO: Not covered by tests
+        "!6" -> "[::1]" -- TODO: Not covered by tests
```

Comment:
See:
https://github.com/PostgREST/postgrest/blob/b2ce322d56687fc67000736876571d8123585502/src/PostgREST/Response/OpenAPI.hs#L387-L393

Let's refactor and make this logic part of the `Network.hs` module.

Reply:
Hm, if the postgREST server resolves !4 to X hostname, then resolving !4 again on the HTTP client would also point to X? 

Can we repeat the process in the client? Then that should work?

Reply:
> I don't think we can reliably repeat the resolution of the special hosts. The conditions could be different at that time.

Yes, agreed. We use `!4` directly in [bindPortTCP](https://hackage.haskell.org/package/streaming-commons-0.2.3.0/docs/Data-Streaming-Network.html#v:bindPortTCP) which gives us the `Socket`. It picks up _any_ hostname and client process can't know it or get that information in any reliable way.

Reply:
> We should just make the client fail with an error when the server is set up like this.

I agree, but our default is `!4`. The error should be clear saying that for this to work you need to override the default and maybe give an example.